src/main/scala/firrtl/RenameMap.scala

```diff
@@ -399,90 +496,114 @@ final class RenameMap private (val underlying: mutable.HashMap[CompleteTarget, S
     if(getCache.contains(key)) {
       getCache(key)
     } else {
-      val getter = recursiveGet(errors)(_)
+
+      // Used to determine if renaming has ended early. This is to prevent the case
+      //   where a fully recursive renaming occurs, e.g.
+      //   Given Renames:
+      //     - (~Top|Top/a:A -> ~Top|A)
+      //     - (~Top|A -> ~Top|B)
+      //   Renaming (~Top|Top/a:A) should return (~Top|A), not (~Top|B)
+      // This is due to our splitting recursive renaming into three parts (top, mid, ast, bot)
+      // We continue renaming even if original == ts.head because if it returns the same target, there are
+      //   still additional renaming of different parts of the target which must occur. If the target actually
+      //   should not be renamed, it must go through all partial renaming parts to actually determine that fact.
+      def continueRenaming[T <: CompleteTarget](original: CompleteTarget, ts: Seq[CompleteTarget]): Boolean =
```

Comment:
I think this is wrong actually, we need to capture the type with a classtag.

```suggestion
      def continueRenaming[T <: CompleteTarget : ClassTag](original: CompleteTarget, ts: Seq[CompleteTarget]): Boolean =
```

Reply:
I'm actually no longer using this function and forgot to delete it. I updated the {reference/instance/module}Get functions to return `Option` instead, so renaming stops when they return `Some`. since renaming can stop on self renames now.